File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseCreateObservability.kt
package com.lextures.android.core.lms

import android.content.Context

/** Lightweight client-only course-create funnel counters (MOB.1; no PII). */
object CourseCreateObservability {
    private const val PREFS = "course_create_observability"

    fun recordStarted(context: Context, mode: String, templateId: String) {
        bump(context, "course_create_started")
        prefs(context).edit()
            .putString("course_create_started.last_mode", mode)
            .putString("course_create_started.last_template", templateId)
            .apply()
    }

    fun recordStepCompleted(context: Context, step: Int) {
        bump(context, "course_create_step_completed")
        prefs(context).edit().putInt("course_create_step_completed.last_step", step).apply()
    }

    fun recordFinished(context: Context, mode: String, templateId: String) {
        bump(context, "course_create_finished")
        prefs(context).edit()
            .putString("course_create_finished.last_mode", mode)
            .putString("course_create_finished.last_template", templateId)
            .apply()
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

    private fun bump(context: Context, key: String) {
        val p = prefs(context)
        p.edit().putInt(key, p.getInt(key, 0) + 1).apply()
    }
}

File: clients/ios/Lextures/Core/LMS/CourseCreateObservability.swift
import Foundation

/// Lightweight client-only course-create funnel counters (MOB.1; no PII).
enum CourseCreateObservability {
    private static let defaults = UserDefaults.standard

    static func recordStarted(mode: String, templateId: String) {
        bump("course_create_started")
        defaults.set(mode, forKey: "course_create_started.last_mode")
        defaults.set(templateId, forKey: "course_create_started.last_template")
    }

    static func recordStepCompleted(step: Int) {
        bump("course_create_step_completed")
        defaults.set(step, forKey: "course_create_step_completed.last_step")
    }

    static func recordFinished(mode: String, templateId: String) {
        bump("course_create_finished")
        defaults.set(mode, forKey: "course_create_finished.last_mode")
        defaults.set(templateId, forKey: "course_create_finished.last_template")
    }

    private static func bump(_ key: String) {
        let prev = defaults.integer(forKey: key)
        defaults.set(prev + 1, forKey: key)
    }
}
